### PR TITLE
Issue 2204 - inaccurate message stats due to nonexistant UTC to local changes

### PR DESCRIPTION
**Pull request by alexjanousekGSA**
This is for issue: [#2204](https://github.com/GSA/notifications-admin/issues/2204)
Related to PR - [2314](https://github.com/GSA/notifications-admin/pull/2314)

## Changes
- Added support for passing a timezone parameter to /service/{id}/statistics/{start}/{days} to properly adjust dates when querying stats.
- Default behavior remains the same (UTC) if no timezone is provided.
- Refactored date calculations for better readability and accuracy, removing the old >= 19 check.
- Updated methods to properly convert timestamps and return data in the requested timezone.
- Added and improved test coverage, including edge cases for daylight savings, timezones and large data queries.

### CODE REVIEW

File: app/dao/date_util.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import calendar
 from datetime import date, datetime, time, timedelta
 
+import pytz
```
Comment (terrazoon) [inline app/dao/date_util.py:4]: I don't think we should be using pytz.  It alternates between being stable and generating flaky weirdness, depending on the upgrade.

Comment (terrazoon) [reply app/dao/date_util.py:4]: we have removed pytz in api when making changes so this would be putting it back.

**Review by terrazoon — CHANGES_REQUESTED** [review]
I don't think there should be any changes on the API side.  It is the job of the front end to pass UTC time to the backend and convert the backend's UTC time to whatever it needs on the front end. 

We do not want to do time conversions on the back end because it adds a lot of complexity and generates defects.   DB time is always UTC, backend expects to receive UTC, backend sends UTC.  Front end does all conversions because only front end knows the local time, etc.

**Comment by ccostino** [comment]
@alexjanousekGSA, thanks for all of the hard work here in trying to make these endpoints reflect the changes needed in the admin to display the data correctly!  There's a lot of good stuff here. 🙂 🎉 

That said, I agree with @terrazoon that these changes should not actually be made in the API itself.  As he noted, the backend and API only deal with UTC for all dates/times:  they're all stored with UTC and they're all processed and returned with UTC.

Any API calls made by the admin (or any other type of client) to the API should be the place where the time conversion happens; either once the data is received, or prior to it going out (though I think the API handles that part at least).

Would it be possible to shift the work you have here to the API client calls and/or views in the admin instead to handle the timezone conversions where needed?  Please let us know if you need help or anything is unclear!

**Comment by alexjanousekGSA** [comment]
> @alexjanousekGSA, thanks for all of the hard work here in trying to make these endpoints reflect the changes needed in the admin to display the data correctly! There's a lot of good stuff here. 🙂 🎉
> 
> That said, I agree with @terrazoon that these changes should not actually be made in the API itself. As he noted, the backend and API only deal with UTC for all dates/times: they're all stored with UTC and they're all processed and returned with UTC.
> 
> Any API calls made by the admin (or any other type of client) to the API should be the place where the time conversion happens; either once the data is received, or prior to it going out (though I think the API handles that part at least).
> 
> Would it be possible to shift the work you have here to the API client calls and/or views in the admin instead to handle the timezone conversions where needed? Please let us know if you need help or anything is unclear!

I appreciate the feedback and understand the preference for keeping the API strictly UTC to stay consistent. I would like to point out that the functionality that were there before did not change whatsoever. There was an enhancement that is a feature, not a bug to account for if a timezone is passed in, the backend can handle a complex conversion. 

In order to keep the logic out of the backend, the front FE would need to perform complicated UTC conversions before sending data to the backend and then convert everything again after receiving the response. That adds extra complexity and risk of defects to the client side. 

If multiple frontend ends ever interact with the API, keeping the time conversion logic in each client could result in duplicate work and inconsistencies. Centralizing complex conversions in the API actually reduce the complexity. 

While I agree that the database should store timestamps in UTC, the API is not a directly interfacing with the database. Its job is to service the clients needs. If the API is already processing and formatting responses to what the FE needs, it makes sense to consider whether time conversion in specific cases is a usability improvement rather an unnecessary complexity. 

The goal here is to reduce complexity in the FE while enhancing the backend api to better serve its clients needs.
